# 单元五作业
    刘易昊 524021910055

## 1.（1分）
为了估计某流行病今年的流行率（即全部人口中总患病的比例），随机采样100人，发现有36个人阳性感染者，那么：

### (1) 该疾病今年的流行率的95%置信区间是多少？

在本样本中，阳性数与阴性数均大于10，不存在极端情况，所以我们选择 Wald 区间

由公式：
$$
\hat{p}=\frac{36}{100}=0.36, n=100
$$

$$
CI=\hat{p}\pm Z_{\alpha/2}\sqrt{\hat{p}(1-\hat{p}/n)}
$$

代入数据

In [2]:
#加载头文件
import numpy as np
import pandas as pd
import scipy.stats as stats
import statistics as sta
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
#中文显示设置
plt.rcParams["font.family"] = ["SimHei", "Microsoft YaHei"]  

In [3]:
def proportion_ci_z(p, n, ci):
    std=np.sqrt(p*(1-p)/n)
    moe=stats.norm.isf((1-ci)/2, 0, 1)*std
    return p-moe, p+moe


l, h=proportion_ci_z(0.36, 100, 0.95)   
print("该疾病今年的流行率的95%置信区间是 : [{:.2f}, {:.2f}]".format(l, h))

该疾病今年的流行率的95%置信区间是 : [0.27, 0.45]


或者直接使用 `proportion` 函数

In [4]:
from statsmodels.stats.proportion import proportion_confint
m, n, alpha=36, 100, 0.05
l, h=proportion_confint(m,n,alpha,method="normal")
print("该疾病今年的流行率的95% Wald 置信区间是 : [{:.2f}, {:.2f}]".format(l, h))


该疾病今年的流行率的95% Wald 置信区间是 : [0.27, 0.45]


### (2) 若该流行病去年的流行率是30%，请用置信区间和z-检验（设显著性水平α=0.05）结果推断今年的流行率是否高于去年？（小数点后保留2位；可能用到的参数Z0.95＝1.64，Z0.975＝1.96)

1. 由（1）可知，今年该疾病流行率的95%置信区间为[0.27,0.45]，去年的流行率为0.3位于该区间内，所以我们有95%的信心认为今年的流行率与去年无显著差别，也就是今年的流行率并不高于去年。

2. 对于右侧z-检验而言，$H_0: p=p_0=0.3, H_1:p>p_0=0.3$，我们构造统计量：
    $$
        z=\frac{\hat{p}-p_0}{\sqrt{p_0(1-P_0)/n}} \overset{H_0}\sim  N(0,1)
    $$
    然后计算p值：

In [5]:
def proportion_z(p_obs, p_0, n):
    obs=(p_obs-p_0)/(np.sqrt(p_0*(1-p_0)/n))
    p=stats.norm.sf(obs, 0, 1)
    return obs, p

z_obs,p=proportion_z(0.36, 0.3, 100)
print('z统计量为', z_obs.round(2))
print('p值为', p.round(2))

z统计量为 1.31
p值为 0.1


同样的，我们也可以使用 `proportion` 函数

In [6]:
from statsmodels.stats.proportion import proportions_ztest
z_obs,p=proportions_ztest(36,100,0.3,alternative="larger",prop_var=0.3)
print('z统计量为', z_obs.round(2))
print('p值为', p.round(2))

z统计量为 1.31
p值为 0.1


由于 $\alpha=0.05, p=0.1>\alpha$，我们不能拒绝 $H_0$，所以在显著性水平为0.05的条件下我们认为该流行病今年和去年的流行率没有差异，即今年的流行率并不高于去年。

### (3) 按APA格式描述和汇报统计结果，并对结果进行解释。

结果汇报：

在本研究中，采用置信区间和z-检验两种方法对某流行病今年、去年的流行率进行检验，结果表明今年该流行病的流行率并不高于去年。该流行病今年流行率95%置信区间为[0.27, 0.45]，去年流行率为0.3包含在置信区间内；计算z-检验统计量 $z=1.31，p=0.1>\alpha$，两种结果均表明该流行病今年的流行率并不高于去年。


## 2.（1分）
某研究小组想比较两个城市居民对某项政策的支持比例是否有差异。从城市A随机抽取了n1=300名居民，其中支持该政策的有x1=180名；从城市B随机抽取了名n2=250名居民，其中支持该政策的有x2=125名。

### (1) 分别计算两个城市居民支持该政策的样本比例p̂1, p̂2及对应的95%置信区间

$\hat{p_1}=\frac{x_1}{n_1}=0.6, \hat{p_2}=\frac{x_2}{n_2}=0.5$

In [7]:
l_a, h_a=proportion_ci_z(0.6, 300, 0.95)   
print("A城市对该政策支持率的95%置信区间 : [{:.2f}, {:.2f}]".format(l_a, h_a))
l_b, h_b=proportion_ci_z(0.5, 250, 0.95)   
print("B城市对该政策支持率的95%置信区间 : [{:.2f}, {:.2f}]".format(l_b, h_b))

A城市对该政策支持率的95%置信区间 : [0.54, 0.66]
B城市对该政策支持率的95%置信区间 : [0.44, 0.56]


或者使用 `proportion` 函数

In [8]:
alpha=0.05
l, h=proportion_confint(180,300,alpha,method="normal")
print("A城市对该政策支持率的95%置信区间 : [{:.2f}, {:.2f}]".format(l, h))
l, h=proportion_confint(125,250,alpha,method="normal")
print("B城市对该政策支持率的95%置信区间 : [{:.2f}, {:.2f}]".format(l, h))

A城市对该政策支持率的95%置信区间 : [0.54, 0.66]
B城市对该政策支持率的95%置信区间 : [0.44, 0.56]


### (2) 计算两个样本比例差值的95%置信区间

对于两个独立样本比例：
$$
\hat{p_1}-\hat{p_2} \sim N(p_1-p_2, \hat{p_1}(1-\hat{p_1})/n_1+\hat{p_2}(1-\hat{p_2})/n_2)
$$

In [9]:
def bi_proportion_ci_z(p_1, p_2, n_1, n_2, ci):
    p=p_1-p_2
    s_1=p_1*(1-p_1)/n_1
    s_2=p_2*(1-p_2)/n_2
    std=np.sqrt(s_1+s_2)
    moe=stats.norm.isf((1-ci)/2, 0, 1)*std
    return p-moe, p+moe

p_1=0.6
p_2=0.5
n_1=300
n_2=250

l, h=bi_proportion_ci_z(p_1, p_2, n_1, n_2, 0.95)  
print("两个样本比例差值的95%置信区间为 : [{:.2f}, {:.2f}]".format(l, h))

两个样本比例差值的95%置信区间为 : [0.02, 0.18]


同样的，我们也可以用 `proportion` 函数进行计算

In [10]:
from statsmodels.stats.proportion import confint_proportions_2indep
l, h= confint_proportions_2indep(180, 300, 125, 250, method='wald')
print("两个样本比例差值的95%置信区间为 : [{:.2f}, {:.2f}]".format(l, h))

两个样本比例差值的95%置信区间为 : [0.02, 0.18]


### (3) 进行双样本比例z-检验（设显著性水平α=0.05），判断两个城市居民对该政策的支持比例是否有显著差异，并按APA格式描述和汇报统计结果。（小数点后保留2位；可能用到的参数Z0.95＝1.64，Z0.975＝1.96)

$H_0:p_1-p_2=0$，$H_1:p_1-p_2 \neq 0$

我们采取合并比例法：$\hat{p}=\frac{x_1+x_2}{n_1+n_2}=0.55$，$ sem=\sqrt{\hat{p}(1-\hat{p})(\frac{1}{n_1}+\frac{1}{n_2})} $

构造z统计量： $ z=\frac{\hat{p_1}-\hat{p_2}}{\sqrt{\hat{p}(1-\hat{p})(\frac{1}{n_1}+\frac{1}{n_2})}} $


In [11]:
def bi_proportion_z(x_1, x_2, n_1, n_2):
    p_1=x_1/n_1
    p_2=x_2/n_2
    p=(x_1+x_2)/(n_1+n_2)
    sem=np.sqrt(p*(1-p)*(1/n_1+1/n_2))
    obs=(p_1-p_2)/sem
    p_val=stats.norm.sf(obs, 0, 1)*2
    return obs, p_val

x_1=180
x_2=125
n_1=300
n_2=250

z_obs, val=bi_proportion_z(x_1, x_2, n_1, n_2)
print('z统计量为', z_obs.round(2))
print('p值为', val.round(2))

z统计量为 2.35
p值为 0.02


也可以用 `proportion` 函数

In [12]:
from statsmodels.stats.proportion import proportions_ztest
count = np.array([180, 125])
nobs = np.array([300, 250])
z_obs,p=proportions_ztest(count, nobs,alternative="two-sided")
print('z统计量为', z_obs.round(2))
print('p值为', p.round(2))

z统计量为 2.35
p值为 0.02


$ \alpha=0.05 $，$ p=.02<\alpha $ 从而拒绝 $H_0$，表明在显著性水平0.05的条件下，我们认为两个城市居民对该政策的支持比例是有差异的。 

**结果汇报：**

独立样本比例检验表明，两个城市居民对该政策的支持率是有显著差异的，A城市随机抽取300人支持率为0.6显著高于B城市随机抽取250人的支持率0.5，两个样本的支持率差异的95%置信区间为[0.02, 0.18]，z统计量为2.35，p值为.02。


## 3.（2分）
为了评估某新型抗病毒药物在不同医院的推广使用是否对甲型流感（甲流）感染率产生影响，研究者在三家医院中，分别对已使用该药物预防的门诊患者进行了流行病学调查，采用分层随机抽样方法：在医院A中随机抽查了 300 人，在医院B中抽查了 400 人，在医院C中抽查了 200 人。检测发现：医院A中 30 人为甲流阳性；医院B中 25 人为阳性；医院C中 10 人为阳性。

### （1）请写出观测频数RC联立表；

In [13]:
data_o = pd.DataFrame(
    [
        [30,25,10],
        [270,375,190],
    ],
    index=["甲流阳性","甲流阴性"],
    columns=["医院A","医院B","医院C"])

data_o

,医院A,医院B,医院C
甲流阳性,30,25,10
甲流阴性,270,375,190


### （2）请写出卡方检验的零假设和备择假设；

零假设$H_0$：某新型抗病毒药物在不同医院的推广使用与甲型流感感染率相互独立

备选假设$H_1$：某新型抗病毒药物在不同医院的推广使用与甲型流感感染率有关联

### （3）请写出期望观测频数RC联立表（四舍五入取整）

根据公式 $ E_{ij}=\frac{行i总数\times列j总数}{N} $，计算的到期望观测频数列联表：


In [14]:
chi2,p,dof,expected=stats.chi2_contingency(data_o)
data_e = pd.DataFrame(
    expected.round(2),
    index=["甲流阳性","甲流阴性"],
    columns=["医院A","医院B","医院C"])

data_e['行总'] = data_e.sum(axis=1)
data_e.loc['列总'] = data_e.sum(axis=0)

data_e

,医院A,医院B,医院C,行总
甲流阳性,21.67,28.89,14.44,65.0
甲流阴性,278.33,371.11,185.56,835.0
列总,300.00,400.00,200.00,900.0


### （4）计算自由度、卡方值与p值（保留小数点后2位）



In [15]:
print('自由度为', dof)
print('卡方值为', chi2.round(2))
print('p值为', p.round(2))

自由度为 2
卡方值为 5.49
p值为 0.06


### （5）计算效应量Cramer’s V值 （保留小数点后两位）

根据公式：
$$
V=\sqrt{\frac{\chi^2}{N\cdot min(r-1,c-1)}}=\sqrt{\frac{5.49}{900\times 1}}=0.0781 \approx 0.08
$$

### （6）按APA格式描述和汇报统计结果 。

卡方检验表明，某新型抗病毒药物在不同医院的推广使用与甲型流感感染率之间存在的关联强度并不显著，其中 $ \chi^2(2, N=900)=5.49 $，$ p=.06>.05 $，$ Cramer's V=0.08 $，表明两者的关联强度可以忽略。

## 4.（2分）
以下是某地区初中随机采样的1000名学生的体重分布表，记录了从40-70kg每5kg区间，以及40kg以下，70kg以上总共8个体重区间的学生数目。请根据这个表格推断这个地区的初中学生体重是否服从均值为55kg，标准差为8kg 正态分布？请写出计算过程和结果，并按APA格式描述和汇报统计结果。

根据题意，本题是完全指定分布的情况，指定分布为 $ N(55, 8^2) $，需要估计的参数为0个，自由度为7。

首先提出假设：
- $H_0$；身高服从正态分布$ N(55, 8^2) $
- $H_1$：身高不服从正态分布$ N(55, 8^2) $

然后进行计算：




In [16]:
# 输入8个区间的观测值、零假设检验的数据
array_o = np.array([32, 72, 158, 236, 232, 162, 76, 32])
n = 1000
mu, sigma = 55, 8
boundaries = [40, 45, 50, 55, 60, 65, 70]

# 计算零假设检验成立下的期望，这里用了个for循环
probs=[]

probs.append(stats.norm.cdf(boundaries[0], mu, sigma))
for i in range(len(boundaries)-1):
    p=stats.norm.cdf(boundaries[i+1], mu, sigma) - stats.norm.cdf(boundaries[i], mu, sigma)
    probs.append(p)
probs.append(1-stats.norm.cdf(boundaries[-1], mu, sigma))

array_e = np.array(probs)*n

# 利用函数，计算卡方和p值
chi2, p= stats.chisquare(f_obs=array_o, f_exp=array_e)

#输出列联表以及最终结果
df_result = pd.DataFrame({
    '体重区间': ['<40', '[40,45)', '[45,50)', '[50,55)', '[55,60)', '[60,65)', '[65,70)', '>70'],
    'Oi': array_o,
    'Ei': array_e.round(2)
})

print(df_result)
print(f"卡方统计量: {chi2:.2f}")
print(f"自由度: {len(array_o) - 1}")
print(f"P值: {p:.4f}")

      体重区间   Oi      Ei
0      <40   32   30.40
1  [40,45)   72   75.25
2  [45,50)  158  160.34
3  [50,55)  236  234.01
4  [55,60)  232  234.01
5  [60,65)  162  160.34
6  [65,70)   76   75.25
7      >70   32   30.40
卡方统计量: 0.40
自由度: 7
P值: 0.9997


卡方拟合优度检验表明，这个地区的初中生体重服从均值为55kg、标准差为8kg的正态分布，$\chi^2(7,N=1000)=0.40$，$p=.9997>0.05$，在显著性水平为0.05的条件下不能拒绝零假设，即认为该地区初中生体重符合题目所述正态分布。